sdk/python/feast/infra/materialization/snowflake_engine.py

```diff
@@ -0,0 +1,459 @@
+
+            if feature_view.batch_source.field_mapping is not None:
+                fv_latest_mapped_values_sql = _run_snowflake_field_mapping(
+                    fv_latest_values_sql, feature_view.batch_source.field_mapping
+                )
+
+            fv_timestamp = feature_view.batch_source.timestamp_field
+
+            if feature_view.batch_source.created_timestamp_column:
+                fv_created_str = (
+                    f',"{feature_view.batch_source.created_timestamp_column}"'
+                )
+            else:
+                fv_created_str = None
+
+            join_keys = [entity.name for entity in feature_view.entity_columns]
+            join_keys_type = [
+                entity.dtype.to_value_type().name
+                for entity in feature_view.entity_columns
+            ]
+
+            entity_names = "ARRAY_CONSTRUCT('" + "', '".join(join_keys) + "')"
+            entity_data = 'ARRAY_CONSTRUCT("' + '", "'.join(join_keys) + '")'
+            entity_types = "ARRAY_CONSTRUCT('" + "', '".join(join_keys_type) + "')"
+
+            project_name = self.repo_config.project
+
+            """
+            Generate the SQL that maps the feature given ValueType to the correct python
+            UDF serialization function.
+            """
+            feature_sql_list = []
+            for feature in feature_view.features:
+                feature_value_type_name = feature.dtype.to_value_type().name
+
+                feature_sql = _convert_value_name_to_snowflake_udf(
+                    feature_value_type_name, project_name
+                )
+
+                if feature_value_type_name == "UNIX_TIMESTAMP":
+                    feature_sql = f'{feature_sql}(DATE_PART(EPOCH_NANOSECOND, "{feature.name}")) AS "{feature.name}"'
+                else:
+                    feature_sql = f'{feature_sql}("{feature.name}") AS "{feature.name}"'
+
+                feature_sql_list.append(feature_sql)
+
+            features_str = ",\n".join(feature_sql_list)
+
+            if self.repo_config.online_store.type == "snowflake.online":
+                serial_func = f"feast_{project_name}_serialize_entity_keys"
+            else:
+                serial_func = f"feast_{project_name}_entity_key_proto_to_string"
+
+            fv_to_proto = f"""
+                SELECT
+                  {serial_func.upper()}({entity_names}, {entity_data}, {entity_types}) AS "entity_key",
+                  {features_str},
+                  "{fv_timestamp}"
+                  {fv_created_str if fv_created_str else ''}
+                FROM (
+                  {fv_latest_mapped_values_sql}
+                )
+            """
+
+            feature_names = [feature.name for feature in feature_view.features]
+            feature_names_str = '", "'.join(feature_names)
+
+            conn_config = snowflake_materialization_config(self.repo_config)
+            if self.repo_config.online_store.type == "snowflake.online":
```

Comment:
maybe factor out the logic for snowflake online store into a separate method? and likewise for the non-online store logic into another?